### PR TITLE
First draft

### PR DESCRIPTION
**Pull request by vicb**
@realiarthur 
This PR is a work in progress to get feedback.
Let me know if you're happy with the updates here and the plan.
Thanks !

I have used TSDX because:
- It's recommended on [typescriptlang.org](https://www.typescriptlang.org/docs/bootstrap),
- It set ups all the tools for linting, testing, generating all of esm, cjs, and .d.ts output.

Questions:
- [ ] I added the src to the npm packages, is this ok ?
- [ ] Is it ok to format the code using prettier (this would go in a separate commit so that review is not unnecessarily complicated)

Done:
- Conversion to typescript - still needs validation but ok for a preview,
- Fixed a bug in `withValue.js` - `disconectedCallback` -> `disconnectedCallback` (nn)
- Fixed a bug in `utils.js` - the `parsed` variable was in the global namespace
- [ ] Fixed a bug in `withForm.js` `handleBlur()` - missing return type. Fixed by removing recursion. To check.

ToDo:
- [ ] regression on the minified size. Was 5.49 kB before this update. Probably related to terser config,
- [ ] add a TS version of the example,
- [ ] serve the example in watch mode,
- [ ] re-enable husky (package.json),
- [ ] add some doc on types.

### CODE REVIEW

**Comment by vicb** [comment]
I have played a little more with that.

My conclusion is that some of the construct do not play well with typescript:
- the [closest polyfill](https://github.com/realiarthur/lite-form/blob/9a7305361d8daa774fa76d97c3c1da1cdb43a852/src/utils.js#L34),
- the fact that some methods take nameOrEvent - [handleBlur](https://github.com/realiarthur/lite-form/blob/9a7305361d8daa774fa76d97c3c1da1cdb43a852/src/withForm.js#L122) and [handleChange](https://github.com/realiarthur/lite-form/blob/9a7305361d8daa774fa76d97c3c1da1cdb43a852/src/withForm.js#L144),
- there is probably no need have the option to pass options to [withForm](https://github.com/realiarthur/lite-form/blob/9a7305361d8daa774fa76d97c3c1da1cdb43a852/src/withForm.js#L16) nor [withField](https://github.com/realiarthur/lite-form/blob/9a7305361d8daa774fa76d97c3c1da1cdb43a852/src/withField.js#L22)

In the end I feel like converting to TS would only make sense with a few breaking changes. So I am wondering if it should be done in this repo (with a new major release) or if it is better done in a fork.

What do you think ?

**Comment by realiarthur** [comment]
- As I said my point about closest polyfill is that builtin function may
works faster. But if there is no other way, we can make a js function

- handleChange and handleBlur is more like inner functions, so we can do
make whatever we want with this. But somehow formik support both typescript
and arguments like nameOrEvent (at least version 1.x.x)

- why do not pass options to withForm? I’m always use withForm with options
like “validationSchema” or “onSubmit”. It looks like standard HOC pattern
and make code more readable. Or i’m wrong?

If you don't want to deal with it further, of course you can fork it and do
it like you want)

**Comment by vicb** [comment]
> - As I said my point about closest polyfill is that builtin function may works faster. But if there is no other way, we can make a js function

The JS function would be actually be the same as the polyfill, just not on the prototype.

> - handleChange and handleBlur is more like inner functions, so we can do make whatever we want with this. But somehow formik support both typescript and arguments like nameOrEvent (at least version 1.x.x)

It's probably possible to make a working typing with some effort.
Is it worth it ? I don't think so. We know which or name or event we will pass so having different function would make things clearer (and simpler) IMO.
On the other it's a breaking change if someone rely on it.

>- why do not pass options to withForm? I’m always use withForm with options like “validationSchema” or “onSubmit”. It looks like standard HOC pattern and make code more readable. Or i’m wrong? The way I use it is to implement `validationSchema` or `onSubmit` in my Lit Element class. 

With TS, you have to use `class MyComponent extends withForm(LitElement) {...}`  to be able to reference `handle...` anyway so passing options as little sense here.

Dropping the options would allow to simplify the logic (i.e. no more need for `withFormExtended` and simplify `this._onSubmit = (onSubmit || this.onSubmit || function () {}).bind(this)`). But if there are use cases no need to remove it.

> If you don't want to deal with it further, of course you can fork it and do it like you want

I think forking should be last resort. It would be better if can discuss/agree on how to evolve this lib.

**Comment by realiarthur** [comment]
I agree with your points #1 and #2, it not make a huge sense. But it really
useful for me to pass options to withForm and withField.

So, if we can make version 2 with this agreement it will be excellent!

**Comment by vicb** [comment]
Great.
Let's recap

For 1) I am not sure if you want to keep the `closest()` code without patching on `Element.proto` or if the code should be dropped and users should explicitly load a polyfill (the latter is a breaking change).

2) methods that accept `nameOrElement` are split in 2 different methods each - this is a breaking change.

3) we keep the ability to pass options to `withForm` and `withField`.

We need to clarify 1 - and make sure my understanding of 2 and 3 is correct.

**Comment by realiarthur** [comment]
For 1 I think you are right, we can drop it, and make users responsible
about ie.

For 2 and 3 yes, all is like you write

Thank you for discuss!

File: README.md [hunk 1]
```diff
@@ -326,8 +326,8 @@ Instead of using config in the HOC, you can put it in the Component class. This
 - touched: `object`
 - isValid: `boolean`
 - handleSubmit: `function`
-- handleChange: `(path || event, value) => {}`. event: `Event: {target: {name || id, value}} || CustomEvent: {detail: {name || id, value}}`
```
Comment (vicb) [inline README.md:329]: need to be updated.

also add the Element.closest polyfill to the doc

File: examples/index.js [hunk 2]
```diff
@@ -1,21 +1,19 @@
...
 import './forms/NativeInputsForm'
 import './forms/WrappedInputsForm'
+import './style.css'
+
 import BaseClass from './forms/BaseClass'
 import NativeElement from './forms/NativeElement'
-
-import './style.css'
+import { html } from 'lit-element'
+import { render } from 'lit-html'
```
Comment (vicb) [inline examples/index.js:12]: I have a VSCode ext to sort imports... sorry for the noise.

File: package.json [hunk 3]
```diff
@@ -2,15 +2,7 @@
...
-  "scripts": {
-    "start": "tsdx watch",
-    "build": "tsdx build",
-    "test": "tsdx test",
-    "lint": "tsdx lint",
-    "prepare": "tsdx build",
-    "size": "size-limit",
-    "analyze": "size-limit --why"
-  },
+  "scripts": {},
```
Comment (vicb) [inline package.json:5]: I dropped tsdx.

File: rollup.config.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import postcss from 'rollup-plugin-postcss';
```
Comment (vicb) [inline rollup.config.js:1]: This is a WIP.

Still need to generate all of umd, esm, cjs (dev and minified versions).

`rollup -cw` works and serve the ex on localhost:3000

File: src/withForm.ts [hunk 5]
```diff
@@ -133,49 +135,31 @@ const withFormExtended = <T extends Constructor<ConnectableElement & FormConfig>
...
 
-    // TODO: Check remove recursion. I think === 'string' was missing a return.
     // Set touched for name and handleValidate if needed
-    handleBlur = (nameOrEvent: string | Event): void => {
-      let name: string;
-
-      // Extracts the name from the event.
-      if (typeof nameOrEvent !== 'string') {        
-        const eventTarget = getEventTarget(nameOrEvent)
+    handleBlur = (event: Event | CustomEvent, name?: string): void => {
```
Comment (vicb) [inline src/withForm.ts:139]: For `handleBlur` and `handleChange` I have added a second parameter to be able to override the name when called from `withField`. It allow for some code simplification while keeping types simple.

File: src/withError.ts [hunk 6]
```diff
@@ -1,18 +1,23 @@
...
+
 import { EVENTS } from './withForm'
-import get from 'lodash.get'
+import get from 'lodash-es/get'
+import { withFormClass } from './withFormClass'
 
 // Super need to have "name" attribute
 // HoC which returns "error" and "touched" from form by "name"
-export const withError = Сomponent => {
-  return withFormClass(
```
Comment (vicb) [inline src/withError.ts:8]: Note the change here:

```typescript
// from
return withFormClass(class extends Component {...});

// to
return class extends withFormClass(Component) {...}
```
I believe those are equivalent but the update is easier to read ?

**Review by vicb — COMMENTED** [review]
Hey,

Could you please take a look at the updates here ?

File: README.md [hunk 7]
```diff
@@ -318,16 +318,16 @@ Instead of using config in the HOC, you can put it in the Component class. This
...
 
 ##### Props
 - values: `object`
 - errors: `object`
 - touched: `object`
 - isValid: `boolean`
 - handleSubmit: `function`
-- handleChange: `(path || event, value) => {}`. event: `Event: {target: {name || id, value}} || CustomEvent: {detail: {name || id, value}}`
-- handleBlur: `(path || event) => {}`. event: `Event: {target: {name || id}} || CustomEvent: {detail: {name || id}}`
+- handleChange: `(event, value) => {}`. event: `Event: {target: {name || id, value}} || CustomEvent: {detail: {name || id, value}}`
```
Comment (realiarthur) [inline README.md:329]: I noticed that new function does not accept "value" argument

File: src/withField.ts [hunk 8]
```diff
@@ -10,63 +12,57 @@ Super need to have "name" or "id" attribute
...
 
-      handleBlur = () => {
-        this._formClass.handleBlur(this.name || this.id)
+      handleBlur = (event: Event | CustomEvent): void => {
+        this._formClass.handleBlur(event, this.name || this.id)
       }
 
-      connectedCallback() {
+      connectedCallback(): void {
+        const {captureBlur, listenChange} = config ?? {};
```
[the remark this answers is not in the document]
Comment (vicb) [reply src/withField.ts:49]: It should probably be provided on l38: `const withFieldExtended = (config: FieldConfig = {...})`

File: src/withField.ts [hunk 9]
```diff
@@ -10,63 +12,57 @@ Super need to have "name" or "id" attribute
...
-          const value = getValueFromEventTarget(eventTarget)
-          changeExecutor(value)
-          return
-        }
-
-        // if value
-        changeExecutor(eventOrValue)
+const withFieldExtended = (config?: FieldConfig) => <T extends Constructor<CustomElement<{name?: string, id?:string}>>>(Component: T) =>
+    class LiteField extends withValue(Component) {
+      handleChange = (event: Event | CustomEvent): void => {
```
[the remark this answers is not in the document]
Comment (vicb) [reply src/withField.ts:40]: See my comments in withForm

File: src/withForm.ts (hunk 5 above)
Comment (vicb) [reply src/withForm.ts:139]: I understand that in JS it's "common" to have a function that has different behavior according to the parameters... because we can.

However I'm not sure if it is the best design.

For example the former `handleChange` would return a setter when called with a name and void if called with an event.

Maybe:

```
setTouched(name: string): void {
  this._touched = set(this._touched, actualName, true)
  if (this._validateOnBlur) {
    this.handleValidate()
  }    
}

handleBlur(event) {
  const eventTarget = getEventTarget(event)
  if (eventTarget) {
    this.setTouched(eventTarget.name);
  }
}
```

This way:
- `handle<EventName>()` does what it says it does: handling the event,
- you can still call `setTouched(...)` if you do not have an event (cf your comments in `withField`)
- would be same way for `handleChange`, extracting the logic to `setValueAndValidate(name, value)`

File: README.md (hunk 7 above)
Comment (vicb) [reply README.md:329]: See my comments in `withForm`.
TL;DR I am not a big fan of function behavior changing according to the type of the params and would prefer 2 different methods.

File: README.md (hunk 1 above)
Comment (vicb) [reply README.md:329]: I knew that this is probably something that need some discussion so I left as is. WIP.

**Review by vicb — COMMENTED** [review]
Thanks for your review.

I have added some suggestion in `withForm` they should also be applicable to `withField`.
That will be breaking changes.
It is also possible to not include those breaking change and keep the former behavior.
Let me know what you prefer.

As noted in the comments the current setup does not bundle lodash (to stay consistent with your setup). Bundling lodash increases the bundle size by a lot. It should be possible to come with implementation of `get`, `set` and `cloneDeep` that are much lighter. I don't think this should be a priority of this CL though. There are probably libs on npm.

File: README.md (hunk 7 above)
Comment (realiarthur) [reply README.md:329]: I'm talking about what is worth removing `value` from arguments here - it is not actual anyway.

File: src/withField.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (realiarthur) [reply src/withField.ts:49]: agree

**Review by realiarthur — COMMENTED** [review]
Thank you!

File: src/withForm.ts (hunk 5 above)
Comment (vicb) [reply src/withForm.ts:139]: Alternate proposal:

1. I think that for consistency `withField.handleBlur()` should call `withField.setTouched(name || id)` that will delegate to `this._formClass.setTouched(...)`

2. May be we `setValue` can be be changed to accept a validation param `.setValue(value, validation = false)`. Also to be consistent `setValue(value, validation)` would be implemented on both `withField` and `withForm`. The former would delegate to the later. 

3. `withField.handleChange(event)` would call `withField.setValue(value, true)` which would delegate to `withForm.setValue(name || id, value, true)`

4. `handle<EventName>(event)` methods are listeners that only take args.

The difference in my proposal is that `setTouched` and `setValue` are also implemented in `withField`. That is mostly for consistency so that `withField` and `withForm` have a similar. 
For now I am only using `withForm` in my code so you have much more experience to decide which of the proposal makes more sense. I'll implement whatever you prefer.

Thanks

Comment (realiarthur) [reply src/withForm.ts:139]: Great idea! Really like it all!

Comment (vicb) [reply src/withForm.ts:139]: Thanks !

I'll work on the changes described here this week.
Once everything looks good, I'll polish the CL (adding cjs, module and umd outputs, updating the doc, ...).

Comment (vicb) [reply src/withForm.ts:139]: Done

File: src/withField.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (vicb) [reply src/withField.ts:49]: done

**Comment by vicb** [comment]
I have implemented the changes we discussed.

I'll update the description of this PR to list the remaining tasks.

File: package.json [hunk 10]
```diff
@@ -2,11 +2,14 @@
   "name": "lite-form",
   "description": "Form HOC for Web Components. It's also works for LitElement.",
   "version": "1.0.2",
```
Comment (realiarthur) [inline package.json:4]: need to increase major version?

File: src/withField.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+       const eventTarget = getEventTarget(event)
+        if (eventTarget) {
+          const value = getValueFromEventTarget(eventTarget)
+          this.setValue(value, true)
+        }
+      }
+
+      setTouched = (name: string): void => {
+        this._formClass.setTouched(name);
```
Comment (realiarthur) [inline src/withField.ts:54]: Think it will be better to get name from class, and not from argument: 

`this._formClass.setTouched(this.name || this.id);`

It make execution of setTouched easier from inputs components

File: src/withForm.ts [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+            set(obj, fieldName, true)
+            return obj
+          },
+          {}
+        )
+      }
+    }
+
+    // Set touched and handleValidate if needed
+    setTouched = (name: string): void => {
```
Comment (realiarthur) [inline src/withForm.ts:139]: Do we need another argument 'validate = false' here, for user can use it without validation. 
Also call it with  setTouched(name, true) from withForm. handleBlur()

File: src/withField.ts (hunk 11 above)
Comment (realiarthur) [reply src/withField.ts:54]: And maybe it will be useful to have argument "validate" here. As in setValue

File: src/withForm.ts (hunk 12 above)
Comment (realiarthur) [reply src/withForm.ts:139]: and 
if (this._validateOnBlur) {
  this.handleValidate()
}   
`
can be a part of withForm.handleBlur() - think it will be more expected behavior

File: src/withForm.ts [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+        this.handleValidate()
+      }
+    }
+
+    // handle Change events
+    handleChange = (event: Event | CustomEvent): void  => {
+      const eventTarget = getEventTarget(event)
+      if (eventTarget) {
+        const value = getValueFromEventTarget(eventTarget)
+        this.setValue(eventTarget.name || eventTarget.id, value, true)
```
Comment (realiarthur) [inline src/withForm.ts:168]: maybe it will be better: 
`this.setValue(eventTarget.name || eventTarget.id, value, this._validateOnChange)`
and remove `this._validateOnChange` from condition in withForm.setValue

**Review by realiarthur — CHANGES_REQUESTED** [review]
Thank you! Please see few my comments

File: package.json (hunk 10 above)
Comment (vicb) [reply package.json:4]: Done

File: src/withField.ts (hunk 11 above)
Comment (vicb) [reply src/withField.ts:54]: Done:
- `this._formClass.setTouched(this.name || this.id);`,
- add a validate argument (default to `false` for consistency).

> Also it will make withForm.handleBlur simpler

I fail to see how, could you expand on why/how ?

File: src/withForm.ts (hunk 12 above)
Comment (vicb) [reply src/withForm.ts:139]: Done.

File: src/withForm.ts (hunk 13 above)
Comment (vicb) [reply src/withForm.ts:168]: Done.
Your suggestions to pass `this._validateOnChange` and `this._validateOnBlur` make complete sense.

**Comment by vicb** [comment]
@realiarthur Thanks for the great feedback.

I have integrated your feedback.

I'll work on the remaining tasks later today.

**Comment by vicb** [comment]
Quick update on this CL: I need to be able to block some time to finish the remaining tasks here. I will probably not be able to dedicate this time before December. Do not expect any big process in the next 2 weeks.
Thanks.

File: src/withField.ts (hunk 11 above)
Comment (realiarthur) [reply src/withField.ts]: > I fail to see how, could you expand on why/how ?

I make a slip here. I mean `withField.handleBlur, but it not make sense too. Your implementation looks exactly as needed!

File: examples/bundle.js [hunk 14]
```diff
@@ -5355,16 +5356,15 @@
...
             };
-        }
-        connectedCallback() {
-            var _a, _b;
             const { onSubmit, initialValues, validationSchema, validateOnChange, validateOnBlur } = config !== null && config !== void 0 ? config : {};
             // take params from HOC argument, or from class (if you build base class, like <lite-form>) or default
             this._onSubmit = (onSubmit || this.onSubmit || function () { }).bind(this);
             this._initialValues = initialValues || this.initialValues || {};
             this._validationSchema = validationSchema || this.validationSchema || {};
             this._validateOnBlur = (_a = validateOnBlur !== null && validateOnBlur !== void 0 ? validateOnBlur : this.validateOnBlur) !== null && _a !== void 0 ? _a : true;
```
Comment (realiarthur) [inline examples/bundle.js:5364]: is `_a` and `_b` still in scope and available here if you transfer their initialisation in constructor?

Comment (vicb) [reply examples/bundle.js:5364]: This is the generated bundle. I expect it to be correct if the source is correct so I haven't checked it.

The reason why I moved init to the constructor is that I changed the _props to readonly and those could only be initialized in the constructor.

**Comment by realiarthur** [comment]
@vicb I plan to merge your PR and finish the remaining (build and documentation) tasks in the "2.0" branch. You do not mind?

**Comment by vicb** [comment]
Oh sorry I completely forgot about this PR.

I'll be out for a week and can give a hand after that.
